File: backend/src/api/routes/activity.py
"""API routes for activity feed, polling, and stats."""

import logging
from typing import Any

from fastapi import APIRouter, HTTPException, Query

from src.api.deps import CurrentUser
from src.models.activity import ActivityCreate
from src.services.activity_feed_service import ActivityFeedService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/activity", tags=["activity"])

VALID_PERIODS = {"1d", "7d", "30d", "day", "week", "month"}


def _get_service() -> ActivityFeedService:
    """Get ActivityFeedService instance."""
    return ActivityFeedService()


@router.get("")
async def get_activity_feed(
    current_user: CurrentUser,
    type: str | None = Query(None, description="Filter by activity type"),
    agent: str | None = Query(None, description="Filter by agent"),
    entity_type: str | None = Query(None, description="Filter by related entity type"),
    entity_id: str | None = Query(None, description="Filter by related entity ID"),
    since: str | None = Query(None, description="ISO timestamp lower bound"),
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(50, ge=1, le=200, description="Items per page"),
) -> dict[str, Any]:
    """Get paginated activity feed with filters.

    Args:
        current_user: Authenticated user.
        type: Filter by activity_type.
        agent: Filter by agent name.
        entity_type: Filter by related_entity_type (lead, goal, contact, company).
        entity_id: Filter by related_entity_id UUID.
        since: ISO datetime lower bound for created_at.
        page: Page number (1-indexed).
        page_size: Number of items per page.

    Returns:
        Dict with items, total, page.
    """
    filters: dict[str, str] = {}
    if type:
        filters["activity_type"] = type
    if agent:
        filters["agent"] = agent
    if entity_type:
        filters["related_entity_type"] = entity_type
    if entity_id:
        filters["related_entity_id"] = entity_id
    if since:
        filters["date_start"] = since

    service = _get_service()
    try:
        result = await service.get_activity_feed(
            user_id=current_user.id,
            filters=filters,
            page=page,
            page_size=page_size,
        )
    except Exception:
        logger.exception("Failed to fetch activity feed")
        raise HTTPException(status_code=500, detail="Failed to fetch activity feed")

    logger.info(
        "Activity feed requested",
        extra={
            "user_id": current_user.id,
            "count": result["total_count"],
            "page": page,
        },
    )

    return {
        "items": result["activities"],
        "total": result["total_count"],
        "page": result["page"],
    }


@router.get("/poll")
async def poll_activity(
    current_user: CurrentUser,
    since: str = Query(..., description="ISO timestamp; returns activities after this"),
) -> dict[str, Any]:
    """Poll for new activities since a timestamp.

    Lightweight endpoint designed for frequent polling (every ~10s).

    Args:
        current_user: Authenticated user.
        since: ISO timestamp; only activities created after this are returned.

    Returns:
        Dict with items list and count.
    """
    service = _get_service()
    try:
        activities = await service.get_real_time_updates(
            user_id=current_user.id,
            since_timestamp=since,
        )
    except Exception:
        logger.exception("Failed to poll activity")
        raise HTTPException(status_code=500, detail="Failed to poll activity")

    return {"items": activities, "count": len(activities)}


@router.get("/stats")
async def get_activity_stats(
    current_user: CurrentUser,
    period: str = Query("7d", description="Period: 1d, 7d, 30d, day, week, month"),
) -> dict[str, Any]:
    """Get activity summary stats for a period.

    Args:
        current_user: Authenticated user.
        period: One of 1d, 7d, 30d, day, week, month.

    Returns:
        Dict with total, by_type, by_agent, period, since.
    """
    if period not in VALID_PERIODS:
        raise HTTPException(
            status_code=422,
            detail=f"Invalid period '{period}'. Must be one of: {', '.join(sorted(VALID_PERIODS))}",
        )

    service = _get_service()
    try:
        stats = await service.get_activity_stats(
            user_id=current_user.id,
            period=period,
        )
    except Exception:
        logger.exception("Failed to fetch activity stats")
        raise HTTPException(status_code=500, detail="Failed to fetch activity stats")

    return stats


@router.get("/{activity_id}")
async def get_activity_detail(
    activity_id: str,
    current_user: CurrentUser,
) -> dict[str, Any]:
    """Get a single activity with full reasoning chain."""
    from src.services.activity_service import ActivityService

    service = ActivityService()
    activity = await service.get_activity_detail(current_user.id, activity_id)
    if activity is None:
        raise HTTPException(status_code=404, detail="Activity not found")
    return activity


@router.post("")
async def record_activity(
    current_user: CurrentUser,
    body: ActivityCreate,
) -> dict[str, Any]:
    """Record a new activity (internal use by agents)."""
    service = _get_service()
    try:
        result = await service.create_activity(
            user_id=current_user.id,
            activity_type=body.activity_type,
            title=body.title,
            description=body.description,
            agent=body.agent,
            reasoning=body.reasoning,
            confidence=body.confidence,
            related_entity_type=body.related_entity_type,
            related_entity_id=body.related_entity_id,
            metadata=body.metadata,
        )
    except Exception:
        logger.exception("Failed to record activity")
        raise HTTPException(status_code=500, detail="Failed to record activity")

    return result

File: frontend/src/api/activity.ts
import { apiClient } from "./client";

// --- Types ---

export interface ActivityItem {
  id: string;
  user_id: string;
  agent: string | null;
  activity_type: string;
  title: string;
  description: string;
  reasoning: string;
  confidence: number;
  related_entity_type: string | null;
  related_entity_id: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface ActivityFeedResponse {
  items: ActivityItem[];
  total: number;
  page: number;
}

export interface ActivityFilters {
  type?: string;
  agent?: string;
  entity_type?: string;
  entity_id?: string;
  since?: string;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface ActivityPollResponse {
  items: ActivityItem[];
  count: number;
}

export interface ActivityStatsResponse {
  total: number;
  by_type: Record<string, number>;
  by_agent: Record<string, number>;
  period: string;
  since: string;
}

export interface AgentStatusItem {
  status: string;
  last_activity: string | null;
  last_activity_type: string | null;
  last_time: string | null;
}

export interface AgentStatusResponse {
  agents: Record<string, AgentStatusItem>;
}

// --- API functions ---

export async function getActivityFeed(
  filters?: ActivityFilters
): Promise<ActivityFeedResponse> {
  const params = new URLSearchParams();
  if (filters?.type) params.append("type", filters.type);
  if (filters?.agent) params.append("agent", filters.agent);
  if (filters?.entity_type) params.append("entity_type", filters.entity_type);
  if (filters?.entity_id) params.append("entity_id", filters.entity_id);
  if (filters?.since) params.append("since", filters.since);
  if (filters?.page) params.append("page", filters.page.toString());
  if (filters?.page_size)
    params.append("page_size", filters.page_size.toString());

  const url = params.toString() ? `/activity?${params}` : "/activity";
  const response = await apiClient.get<ActivityFeedResponse>(url);
  return response.data;
}

export async function pollActivity(
  since: string
): Promise<ActivityPollResponse> {
  const response = await apiClient.get<ActivityPollResponse>(
    `/activity/poll?since=${encodeURIComponent(since)}`
  );
  return response.data;
}

export async function getActivityStats(
  period = "7d"
): Promise<ActivityStatsResponse> {
  const response = await apiClient.get<ActivityStatsResponse>(
    `/activity/stats?period=${encodeURIComponent(period)}`
  );
  return response.data;
}

export async function getAgentStatus(): Promise<AgentStatusResponse> {
  const response = await apiClient.get<AgentStatusResponse>(
    "/activity/agents"
  );
  return response.data;
}

export async function getActivityDetail(
  activityId: string
): Promise<ActivityItem> {
  const response = await apiClient.get<ActivityItem>(
    `/activity/${activityId}`
  );
  return response.data;
}
